controller_manager:
  ros__parameters:
    update_rate: 100 # Hz
    diagnostics:
      threshold:
        controller_manager:
          periodicity:
            mean_error:
              error: 10
              warn: 5
            standard_devation:
              error: 10
              warn: 5

    joint_state_broadcaster:
      type: joint_state_broadcaster/JointStateBroadcaster

    thruster_pid_controller:
      type: pid_controller/PidController
      ros__parameters:
        dof_names: ["surge, sway, heave, roll, pitch, yaw"]
        command_interface: effort
        reference_and_state_interfaces: [velocity]
        gains:
          surge:  {p: 5.0, i: 0.0, d: 2.0, i_clamp_max: 2.0, i_clamp_min: -2.0}
          sway:   {p: 5.0, i: 0.0, d: 2.0, i_clamp_max: 2.0, i_clamp_min: -2.0}
          heave:  {p: 6.0, i: 0.0, d: 3.0, i_clamp_max: 2.0, i_clamp_min: -2.0}
          roll:   {p: 3.0, i: 0.0, d: 1.0, i_clamp_max: 1.0, i_clamp_min: -1.0}
          pitch:  {p: 3.0, i: 0.0, d: 1.0, i_clamp_max: 1.0, i_clamp_min: -1.0}
          yaw:    {p: 3.0, i: 0.0, d: 1.5, i_clamp_max: 1.0, i_clamp_min: -1.0}

    axis_to_command_controller:
      type: rov_controllers/AxisToCommandController
      ros__parameters:
        joints: [
          "front_left_bottom_thruster_joint",
          "front_right_bottom_thruster_joint",
          "front_left_top_thruster_joint",
          "front_right_top_thruster_joint",
          "back_left_bottom_thruster_joint",
          "back_right_bottom_thruster_joint",
          "back_left_top_thruster_joint",
          "back_right_top_thruster_joint"
        ]
        position_offset: [16, 30, 16] # centimeters, [x, y, z]
        rotation_offset: [45, 45]     # degrees, [pitch, yaw]
        forward_max_thrust: 20.0      # Newtons
        reverse_max_thrust: 15.0      # Newtons
        wrench_weights: [1.0, 1.0, 1.0, 1.0, 1.0, 1.0]
        effort_weights: [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0]
        qp_time_limit: 50             # ms

        prim_res_threshold: 1e-2
        dual_res_threshold: 1e-2
        
gamepad_parser:
  ros__parameters:
    deadzone: 0.05
    axes:
      left_stick_x:  {binding: "sway", index: 0}
      left_stick_y:  {binding: "surge", index: 1}
      right_stick_x: {binding: "yaw", index: 3}
      right_stick_y: {binding: "pitch", index: 4}
      left_trigger:  {binding: "", index: 2}
      right_trigger: {binding: "", index: 5}
      dpad_x:        {binding: "roll", index: 6}
      dpad_y:        {binding: "heave", index: 7}
    buttons:
      a:                  {binding: "", index: 0}
      b:                  {binding: "", index: 1}
      x:                  {binding: "", index: 2}
      y:                  {binding: "", index: 3}
      left_bumper:        {binding: "", index: 4}
      right_bumper:       {binding: "", index: 5}
      left_stick_button:  {binding: "", index: 9}
      right_stick_button: {binding: "", index: 10}
      menu_button:        {binding: "", index: 7}
      view_button:        {binding: "", index: 6}
      upload_button:      {binding: "", index: 11}
      xbox_button:        {binding: "", index: 8}
